torchao/quantization/prototype/mixed_precision/scripts/BO_w_sensi_samples.py

```diff
@@ -0,0 +1,381 @@
+
+        quantize_(
+            model.to(device=device),
+            intN_weight_only(n=bit_width, group_size=groupsize),
+            filter_fn_sen,
+        )
+
+# calculate perplexity on wikitext-document, need to support more tasks
+def cal_wikitext_ppl(model, tokenizer, limit=62):
+
+    with torch.no_grad():
+        result = evaluate(
+            HFLM(pretrained=model, tokenizer=tokenizer, batch_size=1),
+            get_task_dict("wikitext"),
+            limit=limit
+        )
+
+    return result["results"]["wikitext"]["word_perplexity,none"]
+
+def cal_model_size(model, fqn_to_config):
+    _sum = 0
+    fqn_cofg_dict = dict()
+
+    for fqn, config in fqn_to_config:
+        bit_width, groupsize = config.split("_")
+        bit_width = int(bit_width)
+        groupsize = int(groupsize)
+        bit_zeropoint = 64
+        bit_scale = 8
+
+        fqn_cofg_dict[fqn] = (bit_width, groupsize, bit_zeropoint, bit_scale)
+    # print(fqn_cofg_dict)
+
+    for name, parameter in model.named_parameters():
+        flag = 0
+        for fqn in fqn_cofg_dict:
+            if fqn in name:
+                flag = 1
+                if "self_attn" in name or "mlp" in name:
+                    _sum += parameter.numel() * fqn_cofg_dict[fqn][
+                        0
+                    ] + parameter.numel() // fqn_cofg_dict[fqn][1] * (
+                        fqn_cofg_dict[fqn][2] + fqn_cofg_dict[fqn][3]
+                    )
+        if flag == 0:
+            _sum += parameter.numel() * 16
+
+    _sum_in_byte = _sum / 8.0
+    _sum_in_GB = _sum_in_byte / (1024**3) / 1.0
+    return _sum_in_GB
+
+# return evaluation results to complete BO trials
+def eval(model, tokenizer, limit, fqn_to_config):
+    return {
+        "cal_PPL": (cal_wikitext_ppl(model, tokenizer, limit), 0.0),
+        "model_size": (cal_model_size(model, fqn_to_config), 0.0),
+    }
+
+def load_model(repo_id, device):
+    tokenizer = AutoTokenizer.from_pretrained(repo_id)
+    model = AutoModelForCausalLM.from_pretrained(repo_id, torch_dtype=torch.bfloat16).to(
+        device=device
+    )
+    return model, tokenizer
+
+def define_parameter_list():
+    # define the search space
+    parameter_choices_list = []
+    for i in [2, 3, 4]:
+        for j in [32, 64]:
+            parameter_choices_list.append(str(i) + "_" + str(j))
+
+    for i in [5, 6, 8]:
+        for j in [32, 64, 128, 256]:
+            parameter_choices_list.append(str(i) + "_" + str(j))
+
+    # define the search space for all layers
+    parameters_list = []
+    # skip the first3 and last2
+    for i in range(3, 30):  
+        parameters_list.append(
+            {
+                "name": f".{i}.",
+                "type": "choice",
+                "value_type": "str",
+                "values": parameter_choices_list,
+                "is_ordered": False,
+                "sort_values": False,
+            }
+        )
+
+    return parameters_list
```

Comment:
So this is rather problematic in that all parameters here are categorical (unordered discrete) parameters.

This means that the underlying surrogate model will not be able to exploit closeness in the parameter space as everything is string encoded. That is, we won't be able to use the information that we expect the model performance and size of a `(2, 32)` setting to be closer to that of a `(3, 32)` setting than to that of a `(8, 32)` setting. This information is completely lost to our surrogate models due to the string encoding. 

Instead, you should just define two parameters for each layer, one `bit_width_{i}` parameter with values `[2, 3, 4, 5, 6, 7, 8]` and one `groupsize_{i}` parameter with values `[32, 64, 128, 256]`.

Reply:
Thanks! updated the code accordingly